### ISSUE TITLE
Feature Request: ocular status

### ISSUE THREAD

**Issue by jeffalo** [theforumhelpers/theforumhelpers.github.io#65]
hello it's me business man jeffalo, i am here to suggest showing ocular status under bios on the user list.

**Is your feature request related to a problem?**
i like ocular status

**Describe alternatives you've considered**
n/a ocular is too original

**Additional context**
https://my-ocular.jeffalo.net/docs

**Comment by leahcimto** [comment]
i also like ocular status

**Comment by leahcimto** [comment]
kinda already mentioned at #31

**Comment by SausageMcSauce** [comment]
Speaking of Ocular, I think Ocular colo(u)rs could be used as well as background colo(u)rs for a member's section on the studio members page.

**Comment by leahcimto** [comment]
> Speaking of Ocular, I think Ocular colo(u)rs could be used as well as background colo(u)rs for a member's section on the studio members page.

In my opinion that would kinda screw up the color scheme and make the page look really rainbow. @Accio1 what are your thoughts on that?

**Comment by FunctionalMetatable** [comment]
made a pull request about this i guess

**Comment by leahcimto** [comment]
> made a pull request about this i guess

/cc @Accio1

**Comment by leahcimto** [comment]
What kind of distinction could be made to separate the status from the bio and username?

**Comment by Accio1** [comment]
@jeffalo is there currently a character limit on ocular statuses? What security measures are in place to prevent XSS?

@leahcimto this applies to what you asked above. If they are short enough, then they could be displayed next to the name with a little bubble around them to indicate that they are different than the username. Perhaps a tooltip could also be added to show where these statuses come from and how you can add/change yours.

**Comment by leahcimto** [comment]
> @jeffalo is there currently a character limit on ocular statuses? What security measures are in place to prevent XSS?

@Accio1 I don't think there is a character limit because I once put in 100,000 digits of pi, but jeffalo (or another ocular admin) normally removes the statuses that are too long. Also idk what security measures are in place but it sounds like something jeffalo would add.

**Comment by leahcimto** [comment]
@Accio1 see https://ocular.jeffalo.net/xss-info

**Comment by Accio1** [comment]
@leahcimto that is for ocular displaying posts, not my-ocular displaying statuses.

**Comment by leahcimto** [comment]
o

**Comment by jeffalo** [comment]
> is there currently a character limit on ocular statuses? What security measures are in place to prevent XSS?

there is a 250 character max-length on ocular statuses. 

> What security measures are in place to prevent XSS?

none. you should set `innerText` instead of `innerHTML`, an example implementation can be seen at https://github.com/ScratchAddons/ScratchAddons/blob/master/addons/my-ocular/discuss.js

**Comment by leahcimto** [comment]
dogsmakemehappy says:

I was waiting to see if I was going to get a github account, but at the moment not. For ocular status colors, maybe make each user their own page with the background being default if no ocular status, or ocular status

https://scratch.mit.edu/users/leahcimto/#comments-129544107

**Comment by FunctionalMetatable** [comment]
> dogsmakemehappy says:
> 
> I was waiting to see if I was going to get a github account, but at the moment not. For ocular status colors, maybe make each user their own page with the background being default if no ocular status, or ocular status
> 
> https://scratch.mit.edu/users/leahcimto/#comments-129544107

maybe like shefwerld user viewer?

**Comment by jeffalo** [comment]
ps. if this is added please refer to statuses as ocular statuses and not my-ocular statuses. the api will remain at my-ocular but users will update their status at ocular.jeffalo.net starting around april.

**Comment by leahcimto** [comment]
Moved to In Progress since things seem to be kind of coming together.

> Speaking of Ocular, I think Ocular colo(u)rs could be used as well as background colo(u)rs for a member's section on the studio members page.

Another reason this wouldn't work so well is that it could screw up the CSS since the text is black and if someone had their color be black it would be impossible to read.

**Comment by FunctionalMetatable** [comment]
> Moved to In Progress since things seem to be kind of coming together.
> 
> > Speaking of Ocular, I think Ocular colo(u)rs could be used as well as background colo(u)rs for a member's section on the studio members page.
> 
> Another reason this wouldn't work so well is that it could screw up the CSS since the text is black and if someone had their color be black it would be impossible to read.
> 
> > > dogsmakemehappy says:
> > > I was waiting to see if I was going to get a github account, but at the moment not. For ocular status colors, maybe make each user their own page with the background being default if no ocular status, or ocular status
> > > https://scratch.mit.edu/users/leahcimto/#comments-129544107
> > 
> > 
> > maybe like shefwerld user viewer?
> 
> Yeah I think that's what they mean. Almost like a page for each user with their status and ocular color.

i **could** try make this

**Comment by Shluffy** [comment]
> Moved to In Progress since things seem to be kind of coming together.
> 
> > Speaking of Ocular, I think Ocular colo(u)rs could be used as well as background colo(u)rs for a member's section on the studio members page.
> 
> Another reason this wouldn't work so well is that it could screw up the CSS since the text is black and if someone had their color be black it would be impossible to read.
> 
> > > dogsmakemehappy says:
> > > I was waiting to see if I was going to get a github account, but at the moment not. For ocular status colors, maybe make each user their own page with the background being default if no ocular status, or ocular status
> > > https://scratch.mit.edu/users/leahcimto/#comments-129544107
> > 
> > 
> > maybe like shefwerld user viewer?
> 
> Yeah I think that's what they mean. Almost like a page for each user with their status and ocular color.

still, there will be a problem with text color, like black on a black background. the option would we to allow the viewer to change the color.

**Comment by jeffalo** [comment]
it should be fairly simple to generate an appropriate text colour for the background. [you can even do it with just css](https://css-tricks.com/switch-font-color-for-different-backgrounds-with-css/). otherwise you can always just display favourite colour as a small dot (like how ocular does it) or use drop shadows.

**Comment by Shluffy** [comment]
> it should be fairly simple to generate an appropriate text colour for the background. [you can even do it with just css](https://css-tricks.com/switch-font-color-for-different-backgrounds-with-css/). otherwise you can always just display favourite colour as a small dot (like how ocular does it) or use drop shadows.

even so, the dot may clash with the background

**Comment by FunctionalMetatable** [comment]
@Accio1 Please can I be assigned?

**Comment by Shluffy** [comment]
> @Accio1 Please can I be assigned?

I was hoping I could...

**Comment by Accio1** [comment]
@FunctionalMetatable @Shluffy Thanks for the discussion! No, I don't think that each user should have their own page. Since ocular statuses are limited to 250 chars, I think it is fine to just display them on the list of Forum Helpers page. I will take a crack at this over the weekend to see how it could be implemented while still looking nice.

**Comment by Accio1** [comment]
@jeffalo, I started working on this and have a question: Is there a rate limit? I don't want to spam your server(s), and 100+ requests will have to be made each time the list of Forum Helpers page is loaded.

**Comment by jeffalo** [comment]
@Accio1 don't worry about spamming my server.. i have cloudflare in front of it, which should cache and prevent too much load on my little server.

i might include a bulk request thing in the future, but for now this should work.

if you really wanted to, maybe you could implement some algorithm for lazy loading the statuses, but honestly i don't care lol. ocular / scratch addons already spams my server a lot and nothing bad has happened yet.

**Comment by Accio1** [comment]
This is what I have so far. What are everyone's thoughts? The border is the color that the individual selected for their ocular status.
![image](https://user-images.githubusercontent.com/48496940/114732252-5d94c300-9d10-11eb-9ebb-663be9741cac.png)

**Comment by FunctionalMetatable** [comment]
> This is what I have so far. What are everyone's thoughts? The border is the color that the individual selected for their ocular status.
> ![image](https://user-images.githubusercontent.com/48496940/114732252-5d94c300-9d10-11eb-9ebb-663be9741cac.png)
> 
> @jeffalo @Shluffy @FunctionalMetatable @leahcimto 

I dont think the prefix "Ocular Status: " will look good, maybe remove that part?

**Comment by Shluffy** [comment]
> This is what I have so far. What are everyone's thoughts? The border is the color that the individual selected for their ocular status.
> ![image](https://user-images.githubusercontent.com/48496940/114732252-5d94c300-9d10-11eb-9ebb-663be9741cac.png)
> 
> @jeffalo @Shluffy @FunctionalMetatable @leahcimto

maybe move the status to under the bio

**Comment by Accio1** [comment]
> > This is what I have so far. What are everyone's thoughts? The border is the color that the individual selected for their ocular status.
> > ![image](https://user-images.githubusercontent.com/48496940/114732252-5d94c300-9d10-11eb-9ebb-663be9741cac.png)
> 
> I dont think the prefix "Ocular Status: " will look good, maybe remove that part?

I wanted to ensure that people who visit the page know why some people have two biographies, and if they don't have two, how they would be able to set their own.



> > This is what I have so far. What are everyone's thoughts? The border is the color that the individual selected for their ocular status.
> > ![image](https://user-images.githubusercontent.com/48496940/114732252-5d94c300-9d10-11eb-9ebb-663be9741cac.png)
> > @jeffalo @Shluffy @FunctionalMetatable @leahcimto
> 
> maybe move the status to under the bio

It is already

**Comment by Shluffy** [comment]
> > > This is what I have so far. What are everyone's thoughts? The border is the color that the individual selected for their ocular status.
> > > ![image](https://user-images.githubusercontent.com/48496940/114732252-5d94c300-9d10-11eb-9ebb-663be9741cac.png)
> > 
> > 
> > I dont think the prefix "Ocular Status: " will look good, maybe remove that part?
> 
> I wanted to ensure that people who visit the page know why some people have two biographies, and if they don't have two, how they would be able to set their own.
> 
> > > This is what I have so far. What are everyone's thoughts? The border is the color that the individual selected for their ocular status.
> > > ![image](https://user-images.githubusercontent.com/48496940/114732252-5d94c300-9d10-11eb-9ebb-663be9741cac.png)
> > > @jeffalo @Shluffy @FunctionalMetatable @leahcimto
> > 
> > 
> > maybe move the status to under the bio
> 
> It is already

i mean on the side of the pic.

**Comment by leahcimto** [comment]
> This is what I have so far. What are everyone's thoughts? The border is the color that the individual selected for their ocular status.
> ![image](https://user-images.githubusercontent.com/48496940/114732252-5d94c300-9d10-11eb-9ebb-663be9741cac.png)
> 

@Accio1 That looks awesome! Just a suggestion, but maybe we could get rid of the ocular status thing and put one of those question mark things next to it that you can hover over that provide info about how to change your ocular status, what it is, etc.

**Comment by leahcimto** [comment]
Like this: https://tutorialzine.com/2014/07/css-inline-help-tips

**Comment by jeffalo** [comment]
@Accio1 very clean. i agree that "Ocular Status: " is not neccesary. perhaps put it as the element title so it shows up when you hover over it.

**Comment by Accio1** [comment]
Alright, I've added the title attribute as @jeffalo suggested. I also added a little bit of info to the top of the page to make it clear that the boxes contain Ocular Statuses. I have also added a 0.25s delay between requesting each status so that it doesn't spam jeffalo's servers. I also switched the JS to use fetch, as that is more modern and easier to work with than XMLRequest. Right now, I just am going to finish up some CSS to make sure spacing is correct and then I will create a PR.
